### PR TITLE
feat: [CDS-43509]: Elastigroup Setup Delegate Changes

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Describe your changes

Delegate changes for Setup step with some required changes in manager as well. Built a very generic framework for easy extension to future tasks on this. More details provided in the jira ticket for this

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39848)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger messagecheck

**Comment by piyush-bhuwalka** [comment]
trigger feature-build

**Comment by arvind-choudhary-h** [comment]
*[`120-ng-manager/config.yml` line 261 at r2](https://reviewable.io/reviews/harness/harness-core/39848#-NGRTD8M-Jg_703bftLo:-NGRTD8M-Jg_703bftLp:b-l6hdjr) ([raw file](https://github.com/harness/harness-core/blob/74d1f5e86ab139f4c7d97bed16c6f8b0c311d5be/120-ng-manager/config.yml#L261)):*
> ```YAML
>   authority: localhost
> 
> shouldConfigureWithPMS: ${SHOULD_CONFIGURE_WITH_PMS:-true}
> ```

revert

File: 950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java [hunk 1]
```diff
@@ -399,6 +399,11 @@ public enum TaskType {
   TERRAFORM_TASK_NG_V2(TaskGroup.TERRAFORM_NG, "Terraform Task NG V2"),
   ELASTIGROUP_SETUP_COMMAND_TASK_NG(TaskGroup.ELASTIGROUP, "Elastigroup Setup Command Task"),
   ELASTIGROUP_STARTUP_SCRIPT_FETCH_RUN_TASK_NG(TaskGroup.ELASTIGROUP, "Elastigroup Startup Script Fetch Run Task"),
+  TERRAFORM_SECRET_CLEANUP_TASK_NG(TaskGroup.TERRAFORM_NG, "Terraform Secret Cleanup Task"),
+  TERRAGRUNT_PLAN_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Plan Task", true),
+  TERRAGRUNT_APPLY_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Apply Task", true),
+  TERRAGRUNT_DESTROY_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Destroy Task", true),
+  TERRAGRUNT_ROLLBACK_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Rollback Task", true),
```
Comment (harshjain12) [inline 950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java:406]: Do not add new task type in middle of the list

Comment (piyush-bhuwalka) [reply 950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java:406]: It's added at the end of the list. I did not get you

Comment (harshjain12) [reply 950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java:406]: image
Your task seems to be  above ELASTIGROUP_PARAMETERS_FETCH_RUN_TASK_NG ?

Comment (piyush-bhuwalka) [reply 950-delegate-tasks-beans/src/main/java/software/wings/beans/TaskType.java:406]: This was previous state -

TERRAGRUNT_ROLLBACK_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Rollback Task", true);  -> not my task


This is now -
TERRAGRUNT_ROLLBACK_TASK_NG(TaskGroup.TERRAGRUNT, "Terragrunt Rollback Task", true),
ELASTIGROUP_PARAMETERS_FETCH_RUN_TASK_NG(TaskGroup.ELASTIGROUP, "Elastigroup Parameters Fetch Task");  

this elastigroup parameters fetch run task is mine. TERRAGRUNT_ROLLBACK_TASK_NG is not my task. It was in develop branch

How it should be otherewise

**Review by filip5rovic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39848)* status: 31 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java` line 134 at r6](https://reviewable.io/reviews/harness/harness-core/39848#-NHPGC4c4mM7IQyAoEdt:-NHPGC4c4mM7IQyAoEdu:b-40fu3w) ([raw file](https://github.com/harness/harness-core/blob/b7ebb1d57f70c466033a24b83f205184f95b0963/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java#L134)):*
> ```Java
>     }
> 
>     return Integer.valueOf(2);
> ```

Why is this fixed to number 2?
___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupStepCommonHelper.java` line 172 at r6](https://reviewable.io/reviews/harness/harness-core/39848#-NHPKrerCQvDCTjJg3CH:-NHPKrerCQvDCTjJg3CI:bvay8ol) ([raw file](https://github.com/harness/harness-core/blob/b7ebb1d57f70c466033a24b83f205184f95b0963/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupStepCommonHelper.java#L172)):*
> ```Java
>       } else if (ManifestStoreType.INLINE.equals(startupScriptOutcome.getStore().getKind())) {
>         logCallback.saveExecutionLog(
>             color(format("%nFetching %s from Inline Store", "startupScript"), LogColor.White, LogWeight.Bold));
> ```

Is `%n` needed at the beginning? Also, is `format` needed here?

**Review by harshjain12 — DISMISSED** [review]
:lgtm:

> Reviewed 5 of 30 files at r1, 1 of 7 files at r2, 5 of 14 files at r3, 5 of 5 files at r5, 7 of 13 files at r6, 7 of 8 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39848#-:-NHPTy4K4FMkvrz-B4IW:bnfp4nl)* status: :shipit: complete! ✓

**Comment by piyush-bhuwalka** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java` line 134 at r6](https://reviewable.io/reviews/harness/harness-core/39848#-NHPGC4c4mM7IQyAoEdt:-NHPbWRmDvM9tKItnnXs:b-csbztc) ([raw file](https://github.com/harness/harness-core/blob/b7ebb1d57f70c466033a24b83f205184f95b0963/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java#L134)):*
<details><summary><i>Previously, filip5rovic (Filip Petrovic) wrote…</i></summary><blockquote>

Why is this fixed to number 2?
</blockquote></details>

Yes, it should not be. Removed

**Comment by piyush-bhuwalka** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupStepCommonHelper.java` line 172 at r6](https://reviewable.io/reviews/harness/harness-core/39848#-NHPKrerCQvDCTjJg3CH:-NHPbb1F5VrwJijKKlb-:ba3dfob) ([raw file](https://github.com/harness/harness-core/blob/b7ebb1d57f70c466033a24b83f205184f95b0963/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupStepCommonHelper.java#L172)):*
<details><summary><i>Previously, filip5rovic (Filip Petrovic) wrote…</i></summary><blockquote>

Is `%n` needed at the beginning? Also, is `format` needed here?
</blockquote></details>

Removed

**Comment by piyush-bhuwalka** [comment]
*[`120-ng-manager/config.yml` line 261 at r2](https://reviewable.io/reviews/harness/harness-core/39848#-NGRTD8M-Jg_703bftLo:-NHPbf-3B_qtjbSQCjEf:bwvyox4) ([raw file](https://github.com/harness/harness-core/blob/74d1f5e86ab139f4c7d97bed16c6f8b0c311d5be/120-ng-manager/config.yml#L261)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

revert
</blockquote></details>

Reverted

**Review by filip5rovic — APPROVED** [review]
> Reviewed 5 of 30 files at r1, 1 of 7 files at r2, 5 of 14 files at r3, 5 of 5 files at r5, 7 of 13 files at r6, 3 of 8 files at r7, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39848)* status: 1. unresolved discussions.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java` line 96 at r9](https://reviewable.io/reviews/harness/harness-core/39848#-NHT2IuR4Qnpu-wZ078K:-NHT2IuSA6YvXMQ-tA_B:bely7du) ([raw file](https://github.com/harness/harness-core/blob/7dd8d7fb7f66c4533b8b54f2c27fa8dd18a49d4e/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java#L96)):*
> ```Java
> 
>     elastigroupNamePrefix = isBlank(elastigroupNamePrefix)
>         ? Misc.normalizeExpression(ServiceVersionConvention.getPrefix(
> ```

This should fail if elastigroupPrefix is blank.

**Comment by arvind-choudhary-h** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java` line 96 at r9](https://reviewable.io/reviews/harness/harness-core/39848#-NHT2IuR4Qnpu-wZ078K:-NHT2Z77E6uJhF6DxbtN:b-84q1ch) ([raw file](https://github.com/harness/harness-core/blob/7dd8d7fb7f66c4533b8b54f2c27fa8dd18a49d4e/125-cd-nextgen/src/main/java/io/harness/cdng/elastigroup/ElastigroupSetupStep.java#L96)):*
<details><summary><i>Previously, arvind-choudhary-h (Arvind Choudhary) wrote…</i></summary><blockquote>

This should fail if elastigroupPrefix is blank.
</blockquote></details>

Ok, this is fixed in next set of fixes.

**Comment by piyush-bhuwalka** [comment]
Merged to a common branch, so closing this PR

**Pull request closed without merging**